### ISSUE TITLE
"Wrong whitelist provided" Error Message

### ISSUE THREAD

**Issue by cwarden45** [COMBINE-lab/salmon#682]
Hi,

I have a public 10xV2 sample that I am reprocessing ([SRR13313130](https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR13313130)).  More specifically, I believe that the sample should be **10X Genomics 5' v1**.

I am using the following command:

```
ID=5309-CT-2
R1=../Reads/5309-CT-2_S01_L005_R1_001.fastq.gz
R2=../Reads/5309-CT-2_S01_L005_R2_001.fastq.gz

TYPE=10xV2
#for MAP, download from UCSC Table Browser, and remove 1st line (and then manually add SARS-COV-2 genes)
MAP=/path/to/SARS_COV_2-hg38_RefSeq_2column.txt
REF=/path/to/SARS_COV_2-hg38_salmon
CBWL=/path/to/737K-august-2016.txt

salmon alevin -l ISR --chromium -1 $R1 -2 $R2 -i $REF -p 4 -o $ID --tgMap $MAP --whitelist $CBWL
```

If I use the V3 barcode list, then I _don't_ get an error message.  However, I got more discordant cell counts between CellRanger and STARsolo.  The Kallisto cell barcode recovery rate was also low (<10%).

So, I thought there error was that different cell barcodes should be used for this sample.  I think that matches, what I saw on this FAQ page from 10x Genomics:

https://kb.10xgenomics.com/hc/en-us/articles/115004506263-What-is-a-barcode-whitelist-

So, I downloaded the [737k-august-2016.txt](https://github.com/10XGenomics/cellranger/blob/master/lib/python/cellranger/barcodes/737K-august-2016.txt) file from 10x, and I am testing using that.

However, I am now receiving the following error message:

```
[2021-07-07 17:07:44.192] [alevinLog] [info] Done importing white-list Barcodes
[2021-07-07 17:07:44.192] [alevinLog] [error] Wrong whitelist provided
Please check https://salmon.readthedocs.io/en/develop/alevin.html#whitelist
```

On the [website linked](https://salmon.readthedocs.io/en/develop/alevin.html#whitelist), I see a note saying "**Not 10x 724k whitelist**".

However, I apologize that I don't think I understand this note (and the exact file name is different).

Should I be doing some differently when running Alevin for this sample?

Thank you very much.

Sincerely,
Charles

**Comment by k3yavi** [comment]
Hey @cwarden45 ,

Thanks for reaching out. A couple of thoughts:
1.) You might have to change some flags for 10x 5', checkout this [post](https://github.com/COMBINE-lab/salmon/discussions/649).
2.) You don't have to necessarily provide the cellular barcode from 10x, there has been a few previous discussions like this as well, please checkout [this](https://github.com/COMBINE-lab/salmon/issues/396) and [this](https://github.com/COMBINE-lab/salmon/issues/645).

**Comment by cwarden45** [comment]
Thank you very much, @k3yavi !  The references to the other discussion is very helpful.

I am testing what happens if I use `-lISF` (instead of `-lISR`).

Thanks again!

**Comment by cwarden45** [comment]
I tested changing the parameters, and I am still getting the same error message:

```
[2021-07-08 16:05:50.979] [alevinLog] [info] Done importing white-list Barcodes
[2021-07-08 16:05:50.979] [alevinLog] [error] Wrong whitelist provided
Please check https://salmon.readthedocs.io/en/develop/alevin.html#whitelist
```

However, I will look into the other information and see if I can understand what is happening.  I can also test not using a white list and see if that changes the number to be something like an order of magnitude different that what I would expect from the other sample.

So, I will post an update when I can run alevin without an error message, and try to give some sense of the results that are quantified.

**Comment by k3yavi** [comment]
Hi @cwarden45 ,

So `--whitelist` flag is not intended to consume the 737k list of CB you are providing, that's why the error. Thanks for the issue in the doc, I updated it. Idea behind the flag is to externally provide a list of high confidence barcodes potentially identified by different tools like Cellranger, and quantify *only* those barcodes. One use case would be to take the output file `barcodes.tsv` of the cellranger and provide it to flag and quantify only those barcodes.

Briefly, alevin doesn't use the 737k list of barcodes at all, it identifies barcodes empirically. If you'd like to quantify using the 737k list of barcodes, checkout the next generation of alevin tool, called [alevin-fry](https://github.com/COMBINE-lab/alevin-fry).

**Comment by cwarden45** [comment]
Thank you very much, @k3yavi !

There was a different 10x white list that worked for a different sample (with a different 10x design), and there is a 3rd dataset that is a BD Rhapsody experiment (where I can't use CellRanger).

However, for this particular sample, I also have the CellRanger results.  So, I can test as you have described, and close the ticket.

If I have additional questions about the different cell counts from the different methods, then I will open a different ticket for that separate topic.

Thanks again!

**Comment by cwarden45** [comment]
FYI - for the sample in question, there are **9,974** "filtered" barcodes and **737,280** "raw" barcodes.

If I take the larger list of "raw" barcodes and I remove the "-1" from each of them, then that is what I am currently testing for Alevin.

**Comment by cwarden45** [comment]
Strictly speaking, if I use the barcodes.tsv (decompressed) files from CellRanger (for the _same sample_), I still get the same error message.

I get the error message below, with or without "-1" at the end of the barcodes:

```
[2021-07-12 15:55:48.717] [alevinLog] [info] Done importing white-list Barcodes
[2021-07-12 15:55:48.717] [alevinLog] [error] Wrong whitelist provided
Please check https://salmon.readthedocs.io/en/develop/alevin.html#whitelist
```

If I **don't** use any white list, then I think that should be one possible resolution (mentioned before).

However, in case others have a similar question, I have at least temporarily re-opened the ticket.

This is the command that I am currently using:

```
ID=5309-CT-2
R1=../Reads/5309-CT-2_S01_L005_R1_001.fastq.gz
R2=../Reads/5309-CT-2_S01_L005_R2_001.fastq.gz

TYPE=10xV2
#for MAP, download from UCSC Table Browser, and remove 1st line (and then manually add SARS-COV-2 genes)
MAP=../../Ref_Generation/SARS_COV_2-hg38_RefSeq_2column.txt
REF=../../Ref_Generation/SARS_COV_2-hg38_salmon
CBWL=../CellRanger/5309-CT-2/outs/raw_feature_bc_matrix/barcodes.tsv

/path/to/salmon-1.5.1_linux_x86_64/bin/salmon alevin -l ISF --chromium -1 $R1 -2 $R2 -i $REF -p 4 -o $ID --tgMap $MAP --whitelist $CBWL
```

**Is there anything that you think I should change, if I wanted to confirm that I can run the analysis with some sort of white list?**

**Comment by k3yavi** [comment]
Hi @cwarden45 ,

I think I see where the confusion is coming from. I never realized cellranger dumps the full list of all possible 737k cellular barcodes in the "raw_feature_bc_matrix" matrix. The typical use case of `--whitelist` flag is when you provide the `9,974` cellular barcodes present in the "filtered" folder of the cellranger output. I was in the impression the "raw_feature_bc_matrix" doesn't outputs the full 737k cellular barcodes. I'll clarify this in the document, thanks again for pointing this out.

**Comment by k3yavi** [comment]
Weird I don't see the "raw" matrix output of the cellranger dumping the 737k list of CB from on of the data from [here](https://support.10xgenomics.com/single-cell-gene-expression/datasets/6.0.1/Targeted_SC3v3_CellPlex_Mouse_Brain_Nuclei_Neuroscience_Multiplex). Are you sure the if you do `wc -l ../CellRanger/5309-CT-2/outs/raw_feature_bc_matrix/barcodes.tsv` you get 737k ?

**Comment by cwarden45** [comment]
Strictly speaking, barcodes.tsv**.gz** files are provided.

I was viewing the file in Notepad++, so I think that number (from the line numbers on the left) should be correct.  I will double-check the next time that I am on my work computer.

I will test using the smaller number of barcodes.  However, unlike the larger file, I think this would be different for every sample.  If part of the reason that I wanted to run Alevin is that I wanted an independent quantification (which takes less time), then that may be a notable limitation.  However, for whatever reason, this seems to only be an issue with the v2 sample (the v3 sample worked fine).

So, I will test that, and I will at least confirm if I see the same error message or not.

**Comment by k3yavi** [comment]
Yep, I think that's right and it is expected that every sample would have different number of cells/cellular barcodes. The general idea is to use a frequency distribution to separate high quality barcodes from low quality post quantification. I'm sorry that you are facing issues with v2, but the error simply means you are providing the full list of 737k barcodes which is not an expected behavior for the `--whitelist` flag and in a typical 10x/ Dropseq based experiment one would expect ~10-12k cells/CB.

**Comment by cwarden45** [comment]
Maybe it is a little off-topic, but here is the code for the v3 sample that ran without generating error messages:

```
ID=5k_pbmc_v3

R1a=../Reads/5k_pbmc_v3_S1_L001_R1_001.fastq.gz
R1b=../Reads/5k_pbmc_v3_S1_L002_R1_001.fastq.gz
R1c=../Reads/5k_pbmc_v3_S1_L003_R1_001.fastq.gz
R1d=../Reads/5k_pbmc_v3_S1_L004_R1_001.fastq.gz

R2a=../Reads/5k_pbmc_v3_S1_L001_R2_001.fastq.gz
R2b=../Reads/5k_pbmc_v3_S1_L002_R2_001.fastq.gz
R2c=../Reads/5k_pbmc_v3_S1_L003_R2_001.fastq.gz
R2d=../Reads/5k_pbmc_v3_S1_L004_R2_001.fastq.gz

TYPE=10xV3
#for MAP, download from UCSC Table Browser, and remove 1st line (and then manually add SARS-COV-2 genes)
MAP=../../Ref_Generation/SARS_COV_2-hg38_RefSeq_2column.txt
REF=../../Ref_Generation/SARS_COV_2-hg38_salmon
CBWL=/net/isi-dcnl/ifs/user_data/Seq/Chromium_data/3M-february-2018.txt

../../Ref_Generation/salmon-1.5.1_linux_x86_64/bin/salmon alevin -l ISR --chromium -1 $R1a $R1b $R1c $R1d -2 $R2a $R2b $R2c $R2d -i $REF -p 4 -o $ID --tgMap $MAP --whitelist $CBWL
```

That v3 white list is even **larger** than the v2 white list:

```
cwarden$ wc -l /net/isi-dcnl/ifs/user_data/Seq/Chromium_data/3M-february-2018.txt
6794880 /net/isi-dcnl/ifs/user_data/Seq/Chromium_data/3M-february-2018.txt
```

So, that is part of why I was confused.

However, before I start running the analysis with the shorter set of barcodes unique for this sample, here are the commands that I believe you were asking about


```
cwarden$ wc -l ../CellRanger/5309-CT-2/outs/raw_feature_bc_matrix/barcodes.tsv
737280 ../CellRanger/5309-CT-2/outs/raw_feature_bc_matrix/barcodes.tsv
cwarden$ wc -l ../CellRanger/5309-CT-2/outs/filtered_feature_bc_matrix/barcodes.tsv
9974 ../CellRanger/5309-CT-2/outs/filtered_feature_bc_matrix/barcodes.tsv
```


I would prefer to have an option where I could potentially conclude the cell count is different than provided by CellRanger.  

However, I will at least check to confirm this solves the problem with the error message that I am seeing (with the much smaller cell barcode list).

**Comment by cwarden45** [comment]
If i use the smaller set of barcodes, then I progress further.  However, I still receive an error message (and there no **quants_mat_rows.txt** file):

```
[2021-07-13 13:59:07.109] [alevinLog] [info] Total 0.00 UMI after deduplicating.
[2021-07-13 13:59:07.109] [alevinLog] [info] Total 0 BiDirected Edges.
[2021-07-13 13:59:07.109] [alevinLog] [info] Total 0 UniDirected Edges.
[2021-07-13 13:59:07.134] [alevinLog] [info] Finished optimizer
/var/spool/slurmd/job3050767/slurm_script: line 23: 10494 Floating point exception../../Ref_Generation/salmon-1.5.1_linux_x86_64/bin/salmon alevin -l ISF --chromium -1 $R1 -2 $R2 -i $REF -p 4 -o $ID --tgMap $MAP --whitelist $CBWL

```

If the barcode is on the opposite read, then I am not sure if I should really be using the reverse or reverse complement (possibly even for the full barcode list)?

However, for the sake of this discussion, I will now test not providing any white list.  If that works, then I will close the ticket again.

**Comment by cwarden45** [comment]
@k3yavi has kindly offered to provide some additional assistance for this issue (outside the thread).

However, I still wanted to keep other in the loop, in case they encountered similar issues.

So, in terms of an intermediate update:

**Setting 1**:

_Command 1_:
`/path/to/salmon alevin -l ISF --chromium -1 $R1 -2 $R2 -i $REF -p 4 -o $ID --tgMap $MAP`
_End of Log 1_:
```
[2021-07-13 20:12:34.651] [alevinLog] [info] Starting white listing of 814 cells
[2021-07-13 20:12:34.651] [alevinLog] [info] Starting to make feature Matrix
[2021-07-13 20:12:34.654] [alevinLog] [info] Done making feature Matrix
[2021-07-13 20:12:35.447] [alevinLog] [info] Finished white listing
[2021-07-13 20:12:36.158] [alevinLog] [info] Finished optimizer
0.0408521	8.9925e-05	0.000114595	636780	18682.9	
0.0290163	6.61624e-05	0.000111685	230922	8010.3	
```
_Size of quants_mat_rows.txt 1_: 814 lines/barcodes

**Setting 2:**:
_Command 1_:
`/path/to/salmon alevin -l ISR --chromium -1 $R1 -2 $R2 -i $REF -p 4 -o $ID --tgMap $MAP`
_End of Log 1_:
```
[2021-07-14 09:51:38.564] [alevinLog] [info] Starting white listing of 814 cells
[2021-07-14 09:51:38.564] [alevinLog] [info] Starting to make feature Matrix
[2021-07-14 09:51:38.566] [alevinLog] [info] Done making feature Matrix
[2021-07-14 09:51:39.347] [alevinLog] [info] Finished white listing
[2021-07-14 09:51:39.541] [alevinLog] [info] Finished optimizer
[2021-07-14 09:51:39.564] [jointLog] [warning] NOTE: Read Lib [[ ../Reads/5309-CT-2_S01_L005_R1_001.fastq.gz, ../Reads/5309-CT-2_S01_L005_R2_001.fastq.gz]] :

Greater than 5% of the fragments disagreed with the provided library type; check the file: 5309-CT-2/lib_format_counts.json for details

0.0408521	8.9925e-05	0.000114595	636780	18682.9	
0.0290163	6.61624e-05	0.000111685	230922	8010.3	
```
_Size of quants_mat_rows.txt 1_:  814 lines/barcodes

Technically, this means that the program ran without generating an error message, but this seems strange to me.  So, I think I would prefer to keep the issue open a little bit longer.

**Comment by cwarden45** [comment]
The issue relates to the error message, so maybe I will close the issue in terms of being able to run the program without generating the error message.

However, I think something still doesn't seem right, and I thought I should make that clear.

<table>
  <tbody>
    <tr>
      <th align="center">Method</th>
      <th align="center">SRR13313130</th>
      <th align="center">10x_pbmc_5k</th>
    </tr>
    <tr>
	    <td align="left">CellRanger</td>
            <td align="center">9,974 cells</td>
            <td align="center">4,956 cells</td>
    </tr>
    <tr>
	    <td align="left">STARsolo</td>
            <td align="center">7,587 cells</br>(Summary.csv)</td>
            <td align="center">4,586 cells/br>(Summary.csv)</td>
    </tr>
    <tr>
	    <td align="left">Alevin</td>
            <td align="center"><b>814 cell barcodes?</b></td>
            <td align="center"> 856,224 cell barcodes</td>
    </tr>
    <tr>
	    <td align="left">Kallisto</td>
            <td align="center">79,254 cells</br>(BUSpaRse)</td>
            <td align="center">47,598 cells</br>(BUSpaRse)</td>
    </tr>
    <tr>

</tbody>
</table>

For Alevin and Kallisto, I am not so worried about the exact values, since I was expecting extra work was needed to estimate a cell count from a distribution of measurements for each cell barcode.  However, the number of cell barcodes should be larger than the number of cells, and that seems to match for everything except Alevin for this sample.

In other words, for sample, I think that this is the command that generates the fewest errors/warnings/notes:

`/path/to/salmon alevin -l ISF --chromium -1 $R1 -2 $R2 -i $REF -p 4 -o $ID --tgMap $MAP`

However, I think the cell barcode count is too small.

**Is there anything else that you would recommend trying?**

**Comment by k3yavi** [comment]
Hi @cwarden45 ,
Thanks for the report, I am guessing you have also tried `--expectCells` or `--forceCells` flags ? If not, I think it's worth trying and you can find more information about them [here](https://github.com/COMBINE-lab/salmon/issues/396).

**Comment by cwarden45** [comment]
I did not previously test that, but I am currently running the analysis with the extra `--expectCells 10000` parameter.

I will post an update.  If all goes well, I will then close the ticket.

Thank you again for your help!

**Comment by cwarden45** [comment]
With the extra `--expectCells 10000` parameter, Alevin finished running without error messages and 10,641 cell barcodes in **quants_mat_rows.txt**.

I am not sure how much of a difference it makes if I expect 8,000 cells versus 10,000 cells.  However, this looks much closer to the expected number, and I am closing the ticket.

Thanks again!
